grader/assignments/thread/syscalls.c

```diff
@@ -2,39 +2,59 @@ void* malloc(unsigned long);
 uint64_t  pthread_create();
 uint64_t  pthread_join(uint64_t* wstatus);
 
-uint64_t sumChilds(uint64_t pid, uint64_t acc) {
-  uint64_t* s;
-
-  s = malloc(8);
+uint64_t* status;
 
+uint64_t sumChilds(uint64_t offset, uint64_t pid, uint64_t acc) {
   if (pid) {
     // this is the parent process/thread
     // => accumulate pid from pthread_create call, pid from pthread_join call and exit status code
-    acc = acc + pid + pthread_join(s);
-    acc = acc + *s;
+    acc = acc + pid + pthread_join(status + offset);
+    acc = acc + *(status + offset);
   }
 
   return acc;
 }
 
+uint64_t get_unique_offset(uint64_t pid1, uint64_t pid2, uint64_t pid3) {
+  uint64_t offset;
+
+  offset = 0;
+
+  if (pid1 != 0)
+    offset = offset + 1;
+  offset = offset * 2;
+  if (pid2 != 0)
+    offset = offset + 1;
+  offset = offset * 2;
+  if (pid3 != 0)
+    offset = offset + 1;
```

Comment:
another way to calculate this would be: 
```
((pid1 != 0) * 2 + (pid2 != 0)) * 2 + (pid3 != 0)
```
whereas the self-documenting version would look like: 
```
uint64_t bit0;
uint64_t bit1;
uint64_t bit2;

bit0 = pid3 != 0;
bit1 = pid2 != 0;
bit2 = pid1 != 0;

return (bit2 * 2 + bit1) * 2 + bit0;
```

Reply:
Since selfie strives for simplicity I chose the version with the higher readability.